### ISSUE TITLE
[CT-1301] Behavior change: `full_refresh: true` config should respect `--full-refresh, -f` flag

### ISSUE THREAD

**Issue by jtcohen6** [dbt-labs/dbt-core#6013]
https://docs.getdbt.com/reference/resource-configs/full_refresh

### Current behavior
When configured, `full_refresh: true|false` means: "always/never full-refresh, and always ignore the --full-refresh flag"

Added way back in https://github.com/dbt-labs/dbt-core/pull/2438 / v0.18

### Proposed change
- `full_refresh: true` should mean "always respect the --full-refresh flag"
- `full_refresh: false` should mean "never full refresh, under any circumstances"

That's trickier for intuition, and harder to explain, but definitely more desirable in terms of the behavior it enables. I'm not sure why you'd always want to full-refresh an `incremental` model; at that point it's really just a `table` materialization.

If this is set at the project level:
```yml
models:
  +full_refresh: false
```
It's not possible to override that setting for one particular model ("actually I want this model to respect the `--full-refresh` flag"). This doesn't work as I'd hoped it might:
```sql
{{ config(materialized = 'incremental', full_refresh = none) }}
```

### Is this a breaking change?

Only for people who currently make use of `full_refresh: true`. Again, I'm not actually sure what the use case for doing that would be!

**Comment by b-per** [comment]
The only reason I could think of people using `full_refresh: true` would be from some `env_var()`, e.g. always doing a full-refresh in Dev (and maybe limiting the number of rows) but keeping it incremental in Prod. This looks very niche though. Most of the time that I need this parameter it is to set it to `false` to prevent FR.

**Comment by b-per** [comment]
I think that it is worth keeping this issue active

**Comment by b-per** [comment]
Commenting to keep it on! I still think that we could improve incremental models handling.

**Comment by jtcohen6** [comment]
@b-per I agree! I think we'd want to find some way to change the default here, without breaking existing users' projects. If you can think of a clever way to do that, without requiring _another_ opt-in config (`full_refresh_config_respect_flag: true`), I'm all ears.

This behavior would then also diverge from `--store-failures` / `store_failures`, where the node-level config always takes precedence over the flag. Should we change that behavior too, to match?

**Comment by andrii0yerko** [comment]
I had faced the same problem.

Did anyone find any workarounds to achieve respect for the `--full-refresh` flag?

I tried things like
```
{{
    config(
        materialized='incremental',
        full_refresh = flags.FULL_REFRESH,
        ...
}}
```

but it doesn't work. Seems that full_refresh cannot be set dynamically for some reasons

**Comment by remeus** [comment]
I agree that it is somehow counter-intuitive to have the yaml config taking precedence over the CLI argument for this very specific flag:

> The full_refresh config allows you to control whether a resource will always or never perform a full-refresh. This config overrides the --full-refresh command-line flag.

When you are sampling the data locally, to ensure a quick deterministic run, you want to make sure that models are always triggered in full refresh.

To do so, you need to add the following to `dbt_project.yml`:

```yml
models:
  +full_refresh: "{{ target.name == 'local' | as_bool }}"
```

The problem is that it will prevent you from running any non-local runs in refresh mode, even when adding the "--full-refresh" flag.

**Comment by b-per** [comment]
> I agree! I think we'd want to find some way to change the default here, without breaking existing users' projects. If you can think of a clever way to do that, without requiring another opt-in config (full_refresh_config_respect_flag: true), I'm all ears.

My feeling is that this particular issue might actually be a good use case for an opt-in config. The current behaviour feels a bit like an oversight from the initial design that is not what people are generally expecting.

Another option I was thinking about was to maybe allow `true`, `false` and `None`. `true` and `false` would stays as-is, but setting it to `None` would follow the default behavior of following the normal full refresh rule and flag.

> This behavior would then also diverge from --store-failures / store_failures, where the node-level config always takes precedence over the flag. Should we change that behavior too, to match?

I don't have strong opinions here and to me our docs are pretty clear about store failures so I wouldn't see the need to change the current behavior here, especially if there was not request raised to do so.
